base/compiler/tfuncs.jl

```diff
@@ -381,6 +381,9 @@ function nfields_tfunc(@nospecialize(x))
             return Const(isdefined(x, :types) ? length(x.types) : length(x.name.names))
         end
     end
+    if isa(x, Union)
+        return tmerge(nfields_tfunc(x.a), nfields_tfunc(x.b))
```

Comment:
Could optimize a bit for larger unions by returning immediately if `nfields_tfunc(x.a)==Int` and only recursing down `nfields_tfunc(x.b)`otherwise (or vice versa). Not sure if worth it.